discovery_engine_core/core/src/mab.rs

```diff
@@ -46,113 +57,100 @@ impl BetaSample for BetaSampler {
 pub(crate) trait Bucket<T> {
     /// Returns the alpha parameter of the beta distribution.
     fn alpha(&self) -> f32;
+
     /// Returns the beta parameter of the beta distribution.
     fn beta(&self) -> f32;
-    /// Returns `true` if the bucket contains no elements.
+
+    /// Checks if the bucket is empty.
     fn is_empty(&self) -> bool;
+
     /// Removes the next best element from this bucket and returns it, or `None` if it is empty.
     fn pop(&mut self) -> Option<T>;
 }
 
-fn pull_arms<B, T>(beta_sampler: &impl BetaSample, buckets: &mut [&mut B]) -> Result<T, Error>
-where
-    B: Bucket<T>,
-{
-    let sample_from_bucket = |bucket: &B| beta_sampler.sample(bucket.alpha(), bucket.beta());
-
-    let mut buckets = buckets.iter_mut();
-
-    let first_bucket = buckets.next().ok_or(Error::NoBucketsToPull)?;
-    let first_sample = sample_from_bucket(first_bucket)?;
-
-    let bucket = buckets
-        .try_fold(
-            (first_sample, first_bucket),
-            |max, bucket| -> Result<_, Error> {
-                let sample = sample_from_bucket(bucket)?;
-                if let Ordering::Greater = nan_safe_f32_cmp(&sample, &max.0) {
-                    Ok((sample, bucket))
+/// Samples the next element from the buckets.
+#[allow(clippy::future_not_send)]
+async fn pull_arms<T>(
+    beta_sampler: &impl BetaSample,
+    buckets: &[&RwLock<impl Bucket<T>>],
```

Comment:
I don't think is logically correct to allow modification to the stacks while pulling. This module should not know anything about async on how the data is stored. I think that the user of the mab need to acquire the writing lock for every stack before calling it. Maybe we need to generalize the signature to `&Deref<impl Bucket<T>` or something similar.

Reply:
this is just the async implementation of the previous sync code (as it took `buckets: &mut [&mut impl Bucket<T>]` before and mutated them). i agree that it would be better to split it in some way, should we change it in this pr or in another?

Reply:
But making it async in this way we allow that different pull can act on data that has been updated in the mean while. Nothing in here should be async. I would revert this somehow already here.